File: tools/skdev/descriptor.py
"""descriptor.py - the parsed introspection model for ``describe``.

The device's ``describe`` verb emits a machine-parseable, line-tagged block that
lets a host configure itself: parameter names + ranges + deck scope, config enums,
query names, and a capability bitmask. This module holds the dataclasses that model
that block and :func:`parse_describe`, which turns the (log-filtered) lines into a
:class:`DeviceDescriptor`.

Wire format (one item per line, terminated by a bare ``end`` line, which the caller
strips before handing lines here):

    descr engine=<name> version=<ver> masked=<0|1>
    query  <name> <deck|global> <kind> [int:label ...]
    param <name> <deck|global> <lo>..<hi>
    config <name> <int>:<label> <int>:<label> ...
    query <name> <deck|global>
    caps 0x<hex>
    end
"""

from dataclasses import dataclass, field


@dataclass
class ParamDesc:
    """A settable/gettable parameter: name, scope, and declared numeric range."""
    name: str
    scope: str            # "deck" | "global"
    lo: float
    hi: float


@dataclass
class ConfigDesc:
    """An enumerated config: name plus a {int: label} value map."""
    name: str
    values: dict          # {int: label}, e.g. {0:"slice", 1:"reel", 2:"drift"}


@dataclass
class QueryDesc:
    """A readable state item: platform or engine, indistinguishable on the wire by design."""
    name: str
    scope: str            # "deck" | "global"
    kind: str = "text"    # "bool" | "int" | "float" | "enum" | "text"
    values: dict = field(default_factory=dict)   # Enum only: {int: label}


@dataclass
class DeviceDescriptor:
    """The full parsed control surface reported by ``describe``."""
    engine: str = ""
    version: str = ""
    masked: bool = False   # engine narrowed its live_params/live_configs (see parse_describe)
    params: dict = field(default_factory=dict)    # name -> ParamDesc
    configs: dict = field(default_factory=dict)   # name -> ConfigDesc
    queries: dict = field(default_factory=dict)   # name -> QueryDesc
    caps: int = 0


def parse_describe(lines):
    """Parse a describe block into a :class:`DeviceDescriptor`.

    ``lines`` is the block with log lines already filtered out and ending
    before/at ``end`` (the terminator is not required to be present). Blank
    lines and any unrecognized tags are ignored so a forward-compatible device
    that adds a new tag does not break an older host.
    """
    d = DeviceDescriptor()
    for ln in lines:
        tok = ln.split()
        if not tok:
            continue
        if tok[0] == "descr":
            kv = dict(t.split("=", 1) for t in tok[1:] if "=" in t)
            d.engine, d.version = kv.get("engine", ""), kv.get("version", "")
            # masked=1 means the engine declared which ids it actually implements. With masked=0 the
            # descriptor is the whole ParamId enum and a round-trip sweep proves nothing, so the
            # generic test skips rather than reporting a wall of false failures.
            d.masked = kv.get("masked", "0") == "1"
        elif tok[0] == "param":                   # param <name> <scope> <lo>..<hi>
            name, scope, rng = tok[1], tok[2], tok[3]
            lo, hi = (float(x) for x in rng.split(".."))
            d.params[name] = ParamDesc(name, scope, lo, hi)
        elif tok[0] == "config":                  # config <name> [scope] i:label i:label ...
            # Only ``int:label`` tokens are values; skip any optional scope token
            # (the dispatch doc shows a scope, the tools sketch omits it - accept both).
            vals = {int(k): v for k, v in (p.split(":", 1) for p in tok[2:] if ":" in p)}
            d.configs[tok[1]] = ConfigDesc(tok[1], vals)
        elif tok[0] == "query":       # query <name> <scope> [kind] [i:label ...]
            # Scope tells a caller whether to pass a deck; kind tells it how to parse the reply.
            # Both are optional so older firmware (name+scope, or name alone) still parses.
            scope = tok[2] if len(tok) > 2 else "global"
            kind = tok[3] if len(tok) > 3 else "text"
            vals = {int(k): v for k, v in (p.split(":", 1) for p in tok[4:] if ":" in p)}
            d.queries[tok[1]] = QueryDesc(tok[1], scope, kind, vals)
        elif tok[0] == "caps":                    # caps 0x....
            d.caps = int(tok[1], 16)
    return d

File: web/js/terminal/descriptor.js
// descriptor.js - the parsed introspection model for `describe`.
//
// A port of tools/skdev/descriptor.py. This is what makes a web terminal worth more than a serial
// monitor: `describe` reports the engine's own control surface - every parameter with its real range
// and deck scope, every config with its enum labels, the query vocabulary, and a capability mask - so
// the UI can be GENERATED rather than hard-coded. A slider per advertised param, buttons for the pads
// the engine implements, and nothing at all for the enum entries this engine ignores.
//
// Wire format (one item per line, terminated by a bare `end`, which the caller strips):
//
//     descr engine=<name> version=<ver> masked=<0|1>
//     param  <name> <deck|global> <lo>..<hi>
//     config <name> [scope] <int>:<label> ...
//     query  <name> <deck|global> [kind] [int:label ...]
//     caps   0x<hex>
//     end
//
// Unknown tags are ignored rather than rejected, so firmware that adds a line type does not break an
// older page.

/** @typedef {{name: string, scope: string, lo: number, hi: number}} ParamDesc */
/** @typedef {{name: string, values: Map<number,string>}} ConfigDesc */
/** @typedef {{name: string, scope: string, kind: string, values: Map<number,string>}} QueryDesc */

/**
 * Parse a describe block.
 * @param {string[]} lines  log lines already filtered out, `end` not required to be present
 */
export function parseDescribe(lines) {
  const d = {
    engine: '',
    version: '',
    // masked=1 means the engine declared which ids it actually implements. With masked=0 the
    // descriptor is the whole ParamId enum, so a generated surface would show controls the engine
    // ignores - worth flagging in the UI rather than silently rendering dead sliders.
    masked: false,
    /** @type {Map<string,ParamDesc>} */ params: new Map(),
    /** @type {Map<string,ConfigDesc>} */ configs: new Map(),
    /** @type {Map<string,QueryDesc>} */ queries: new Map(),
    caps: 0,
  };
  for (const line of lines) {
    const tok = line.trim().split(/\s+/).filter(Boolean);
    if (!tok.length) continue;
    switch (tok[0]) {
      case 'descr': {
        const kv = new Map(tok.slice(1).filter((t) => t.includes('=')).map((t) => {
          const i = t.indexOf('=');
          return [t.slice(0, i), t.slice(i + 1)];
        }));
        d.engine = kv.get('engine') || '';
        d.version = kv.get('version') || '';
        d.masked = kv.get('masked') === '1';
        break;
      }
      case 'param': { // param <name> <scope> <lo>..<hi>
        if (tok.length < 4) break;
        const [lo, hi] = tok[3].split('..').map(Number);
        d.params.set(tok[1], { name: tok[1], scope: tok[2], lo, hi });
        break;
      }
      case 'config': { // config <name> [scope] i:label i:label ...
        if (tok.length < 2) break;
        d.configs.set(tok[1], { name: tok[1], values: enumValues(tok.slice(2)) });
        break;
      }
      case 'query': { // query <name> <scope> [kind] [i:label ...]
        if (tok.length < 2) break;
        d.queries.set(tok[1], {
          name: tok[1],
          scope: tok.length > 2 ? tok[2] : 'global',
          kind: tok.length > 3 ? tok[3] : 'text',
          values: enumValues(tok.slice(4)),
        });
        break;
      }
      case 'caps':
        if (tok.length > 1) d.caps = parseInt(tok[1], 16) || 0;
        break;
      default:
        break; // forward-compatible: a tag this page does not know is not an error
    }
  }
  return d;
}

/** `0:slice 1:reel 2:drift` -> Map. Non-`int:label` tokens (an optional scope) are skipped. */
function enumValues(tokens) {
  const out = new Map();
  for (const t of tokens) {
    const i = t.indexOf(':');
    if (i < 0) continue;
    const k = Number(t.slice(0, i));
    if (Number.isInteger(k)) out.set(k, t.slice(i + 1));
  }
  return out;
}

/**
 * The command vocabulary a descriptor implies, for console completion. Mirrors what tools/skterm.py
 * completes against, so the two front-ends offer the same words.
 */
export function vocabulary(desc) {
  const words = new Set([
    'set', 'get', 'param', 'config', 'query', 'cv', 'gate', 'midi', 'pad', 'seq', 'fx',
    'reset', 'preset', 'caps', 'mode', 'describe', 'help', 'note', 'msg', 'transport',
    'test', 'run', 'save', 'load', 'play', 'rec', 'stop', 'clear', 'trig', 'arm', 'disarm',
    'flux', 'grit', 'lock', 'gritmode', 'voct', 'mix', 'size', 'xfade', 'cpu', 'A', 'B',
  ]);
  for (const k of desc.params.keys()) words.add(k);
  for (const k of desc.configs.keys()) words.add(k);
  for (const k of desc.queries.keys()) words.add(k);
  return [...words].sort();
}

/**
 * Parse the `query usb` reply - a flag soup like
 * `boot=1 region=2 clkcfg=1 hsi48=1 usbsel=0 usb33den=1 usb33rdy=1 phy=1 pullup=1` - into ordered
 * pairs, so the UI can render the UsbDiag bring-up snapshot as a table rather than a wall of text.
 */
export function parseUsbDiag(reply) {
  return reply.trim().split(/\s+/).filter((t) => t.includes('=')).map((t) => {
    const i = t.indexOf('=');
    return { key: t.slice(0, i), value: t.slice(i + 1) };
  });
}
